plugins/modules/vmware_guest.py

```diff
@@ -1619,29 +1627,34 @@ def configure_hardware_params(self, vm_obj):
         """
         max_connections = self.params['hardware']['max_connections']
         if max_connections is not None:
-            self.configspec.maxMksConnections = max_connections
-            if vm_obj is None or self.configspec.maxMksConnections != vm_obj.config.maxMksConnections:
+            if vm_obj is None or max_connections != vm_obj.config.maxMksConnections:
                 self.change_detected = True
+                self.configspec.maxMksConnections = max_connections
 
         nested_virt = self.params['hardware']['nested_virt']
         if nested_virt is not None:
-            self.configspec.nestedHVEnabled = nested_virt
-            if vm_obj is None or self.configspec.nestedHVEnabled != bool(vm_obj.config.nestedHVEnabled):
+            if vm_obj is None or nested_virt != bool(vm_obj.config.nestedHVEnabled):
                 self.change_detected = True
+                self.configspec.nestedHVEnabled = nested_virt
 
         temp_version = self.params['hardware']['version']
         if temp_version is not None:
             if temp_version.lower() == 'latest':
                 # Check is to make sure vm_obj is not of type template
                 if vm_obj and not vm_obj.config.template:
-                    try:
+                    config_option_descriptors = vm_obj.environmentBrowser.QueryConfigOptionDescriptor()
+                    available_hw_versions = [int(option_desc.key.split("-")[1]) for option_desc in config_option_descriptors if option_desc.upgradeSupported]
+                    max_hw_version = max(available_hw_versions)
+                    if max_hw_version > int(vm_obj.config.version.split("-")[1]):
+                        self.change_detected = True
+                        self.tracked_changes['hardware.version'] = 'latest'
+                        self.change_applied = True
+                        self.configspec.version = "vmx-%02d" % max_hw_version
+                    if not self.module.check_mode:
                         task = vm_obj.UpgradeVM_Task()
                         self.wait_for_task(task)
                         if task.info.state == 'error':
                             return {'changed': self.change_applied, 'failed': True, 'msg': task.info.error.msg, 'op': 'upgrade'}
-                    except vim.fault.AlreadyUpgraded:
-                        # Don't fail if VM is already upgraded.
-                        pass
             else:
                 try:
                     temp_version = int(temp_version)
```

Comment:
Cool. Since you have the max_hw_version now, could we just set temp_version = max_hw_version when temp_version == 'latest' and then fall through to the existing logic for new_version below, so we wouldn't need any of the logic above for calling the upgrade task?

Reply:
Good idea! I wasn't able to test [this](https://github.com/ansible-collections/community.vmware/pull/1553/commits/939dccb767c55f7bb4462d00f0fb98fe780b7779) yet, but I hope it will work.

Reply:
looks good